Define a single source of version information

gdemonet opened the pull request:
**Component**: build, salt, kubernetes, tests

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**:

Version information is currently hardcoded in multiple places (buildchain, Salt formulas, docs), making it difficult when some of them need to be changed.

**Summary**:

- New single source in `buildchain/buildchain/versions.py`
- Generated `salt/metalk8s/versions.json` containing versions for packages and container images
- Adapted macros for `pkg_installed` and `build_image_name`
- New test scenario, to verify the running version of the cluster matches configuration
- Small clean-up of the Docker client usage in the buildchain (see #947)
- [Not yet included in this draft PR] Fix of `metalk8s_package_manager` (see #1455)

**Acceptance criteria**: 

- Everything should keep working as expected
- Integration branches should only need to change the `buildchain/buildchain/versions.py` file

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #640
Closes: #947 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

gdemonet committed:
Store all image and package versions in buildchain

We define a single source of version information for the whole project,
and use Python to define such a file to easily read it from other
locations, such as tests or docs.

This module will be used throughout the buildchain, but should never
contain anything build-specific (such information can be located in the
relevant parts of other buildchain modules).

Issue: #640

gdemonet committed:
Pull images by version and check digest

Previously, we pulled by digest and applied the declared version as a
tag. This implied that updating the declared version only changed the
tag of the image, without actually changing the image itself.

We now make sure to pull the desired version with both Docker and
Skopeo, checking the SHA256 digest of the result against what we specify
in `buildchain.versions`.

Issue: #640

gdemonet committed:
Use new `versions` module for container images

We introduce a number of helper methods to generate the required
targets, hence losing the declarative quality of the previous approach.

However, the `versions` module provides a simpler, more readable
equivalent.

Issue: #640

gdemonet committed:
Use new `versions` module for packages

This allows for a complete removal of the `packages/packages.list` file.

Issue: #640

gdemonet committed:
Generate versions list in Salt tree

In order to read the list of versions defined in the buildchain from
Salt formulas, we need to render them into a static file (we chose
JSON, since YAML requires adding a dependency, and both are the easiest
to read from).

This is then read by `map.jinja` as an extension to the existing
`defaults.yaml`.

Issue: #640

gdemonet committed:
Only download packages if versions changed

This will prevent the buildchain from always downloading packages if no
version changed since the previous execution.
The use of `doit.tools.config_changed` allows us to emulate the previous
behaviour of having a `file_dep` on `packages/packages.list`.

Issue: #640

gdemonet committed:
Ignore epoch if asking for a version to install

When installing packages with the `pkg_installed` macro, we found that
packages having declared an "Epoch" always fail to install (because we
don't specify it in `versions.json`).
Since we do not care about differences between epochs, we just pass the
accurate flag to ignore such differences.

Issue: #640

gdemonet committed:
Ensure SaltStack repository has expected version

We perform a simple templating of the SaltStack Yum repository
definition file within the `metalk8s-build` container image.
The Salt version comes from `buildchain.versions.SALT_VERSION`.

Issue: #640

gdemonet committed:
Read from versions listing in Salt formulas

We adjust the `pkg_installed` and `build_image_name` macros to make
use of the version information we retrieved from the generated
`versions.json` file.

Issue: #640

gdemonet committed:
Use new `build_image_name` macro

The adapted macro allows to remove all previous occurences of hardcoded
versions for container images.

Issue: #640

gdemonet committed:
Use versioned packages in preflight states

Slight refactor of the installation states for packages in
`preflight.[recommended|mandatory]`.

Note: previous usage with `{{ pkg_installed() }}` is not functional
anymore, one always need to pass a package name to the macro.

Issue: #640

gdemonet committed:
Test the cluster version matches configuration

We introduce a test to verify the running Kubernetes version against our
list of versions from the buildchain.

Note: we create a symlink, under `tests/versions.py`, pointing to
`buildchain/buildchain/versions.py`, for ease of importation within
tests.

Issue: #640

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/640-single-source-of-image-versions` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `improvement/640-single-source-of-image-versions`
* Rebase `improvement/640-single-source-of-image-versions` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/640-single-source-of-image-versions` with contents from `improvement/640-single-source-of-image-versions`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/640-single-source-of-image-versions origin/development/2.1
 $ git merge origin/improvement/640-single-source-of-image-versions
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/640-single-source-of-image-versions
```

TeddyAndrieux commented on salt/metalk8s/macro.sls:
```diff
@@ -1,14 +1,19 @@
     {%- endif %}
     - reload_modules: True
+  {%- else %}
+  {{ raise('Missing information for package "' ~ name ~ '"') }}
```
Maybe better to use do for that
```suggestion
  {%- do raise('Missing information for package "' ~ name ~ '"') %}
```

gdemonet replied:
No no, this usage is documented: https://docs.saltstack.com/en/latest/topics/jinja/index.html#errors

TeddyAndrieux replied:
Oh ! Ok

gdemonet committed:
Fix package manager listing of requires

The previous implementation had the following issues:

- a bug when listing from a repository containing multiple versions of a
  single package (can happen with dirty ISO tree during development),
  as it was not able to choose the 'good one'.
- bad naming of functions/variables and imperfect documentation, leading
  to misunderstandings w.r.t. the goal of this method

We fix the issues by:

- implement filtering based on the `repo:packages` pillar information,
  which is already used by the `pkg_installed` macro
- renaming and documenting these custom methods for easier maintenance

Fixes: #1455

gdemonet committed:
Copy VERSION file to Bastion

This file is read from the `versions` Python module, so we need to copy
it along with the test sources.
Note that we squash previous `scp` occurences into a single `tar` copy
over SSH, which allows us to preserve symlinks.

Issue: #640

slaperche-scality commented on .pylintrc:
```diff
@@ -42,11 +42,12 @@ confidence=
+        bad-continuation,    # I kindly disagree :)
+        locally-disabled,    # Sometime we need to disable a warning locally.
+        suppressed-message,  # Don't pollute the output with useless info.
+        useless-return       # Disagreement with mypy, see https://github.com/python/mypy/issues/3974
```
Hum, I disagree with that change.

I think for the `Optional` case it doesn't hurt to be explicit about returning `None`/
If you took the time to annotate the code with "I will return a value OR `None`", then take the time to return the `None`.

And as GvR commented in the issue: it's about the style recommended in PEP8:

> Be consistent in return statements. Either all return statements in a function should return an expression, or none of them should. If any return statement returns an expression, any return statements where no value is returned should explicitly state this as return None, and an explicit return statement should be present at the end of the function (if reachable).

gdemonet replied:
Well, I agree with you and Guido, but `pylint` doesn't (complaining with `useless-return`), hence the disable here.

slaperche-scality replied:
Ho right, I went to read the issue and thought we wanted to get rid of a warning about "missing" return, not about "useless" one.

Then it's good of course >_<

That being said we should probably link this issue instead: https://github.com/PyCQA/pylint/issues/2300
But it seems to be fixed…

On which code did you had the warning?
Either Pylint was right (maybe you had an if/else with a return in both branches already, so the last return was really useless) or we may have a minimal example to reopen the bug.

slaperche-scality commented on buildchain/buildchain/constants.py:
```diff
@@ -17,11 +17,13 @@
-COREOS_REGISTRY : str = 'quay.io/coreos'
-PROMETHEUS_REGISTRY : str = 'quay.io/prometheus'
-GRAFANA_REGISTRY : str = 'docker.io/grafana'
+CALICO_REPOSITORY     : str = 'quay.io/calico'
```
minor: since you are modifying each line, you could use the occasion to order them by alphabetical order.

gdemonet replied:
Oops I misplaced `INGRESS_...`

slaperche-scality commented on buildchain/buildchain/docker_command.py:
```diff
@@ -141,13 +141,14 @@ def __init__(
     @task_error(docker.errors.APIError)
     def __call__(self) -> Optional[TaskError]:
-        return DOCKER_CLIENT.images.build(
+        DOCKER_CLIENT.images.build(
```
I don't understand the relation between "this is a patch about renaming" and "let's change the return value of the code"?

gdemonet replied:
I roughly stashed all my changes to `docker_command` in this commit :/ I'll try to split it

slaperche-scality commented on buildchain/buildchain/targets/remote_image.py:
```diff
@@ -62,39 +61,18 @@ def __init__(
-        )
+    def remote_fullname(self) -> str:
+        """Complete image name retrieved from the remote repository."""
+        return "{i.repository}/{i._remote_name}:{i.version}".format(i=self)
```
`i`? Could use `img` at least.

Same remarks for the occurrences below.

gdemonet replied:
Yeaaahhh, line length made `i` more appealing, but w/e

slaperche-scality commented on buildchain/buildchain/targets/remote_image.py:
```diff
@@ -150,12 +131,14 @@ def _skopeo_copy(self) -> List[str]:
     def _skopeo_dest(self) -> str:
         """Return the destination, formatted for skopeo copy."""
         if self._use_tar:
+            # NOTE: unused, since `docker-archive` format doesn't work with
```
Ha good catch, we forgot to remove this code when we reverted back to using Docker.

I think the code can be removed in a dedicated commit (with your comment as commit message, and maybe a link to the related issues).

slaperche-scality commented on buildchain/buildchain/image.py:
```diff
@@ -89,244 +90,127 @@ def show() -> str:
+    try:
+        return versions.CONTAINER_IMAGES_MAP[name]
+    except KeyError:
+        raise Exception(
```
Could be a `ValueError` (the `name` was wrong) I guess.

slaperche-scality commented on buildchain/buildchain/image.py:
```diff
@@ -89,244 +90,127 @@ def show() -> str:
 
-NGINX_IMAGE_VERSION : str = '1.15.8'
-NODE_IMAGE_VERSION : str = '10.16.0'
+def _remote_image(name: str, repository: str, **overrides: Any) \
```
I prefer this wrapping:

```python
def _remote_image(
    name: str, repository: str, **overrides: Any
) -> targets.RemoteImage:
```

It avoids the trailing `\` and is consistent with the layout used elsewhere in the build chain.

slaperche-scality commented on buildchain/buildchain/image.py:
```diff
@@ -89,244 +90,127 @@ def show() -> str:
     ]
     return task
 
+# Helpers
```
Some basic docstrings would be good for those helpers.

minor: also I guess the `Helpers`, `Container images to pull` (`List of` is superfluous) and `Container images to build` could be folds.

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -142,52 +139,75 @@ def task__build_repositories() -> Iterator[types.TaskDict]:
+    try:
+        pkg_info = versions.PACKAGES_MAP[name]
+    except KeyError:
+        raise Exception(
```
Same remark as in previous commit: I think a `ValueError` could make sense here and be more specific than `Exception`.

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -142,52 +139,75 @@ def task__build_repositories() -> Iterator[types.TaskDict]:
+            'Missing version for package "{}"'.format(name)
+        )
+
+    try:
```
Could this `try`/`except` be replaced by: `build_id_str, _, _ = pkg_info.release.partition('.')`?

e.g.:

```
>>> no_dot = 'foo'
>>> one_dot = 'foo.bar'
>>> more_dot = 'foo.bar.baz'
>>> res, _ = no_dot.split('.', 1); print(res)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: not enough values to unpack (expected 2, got 1)
>>> res, _ = one_dot.split('.', 1); print(res)
foo
>>> res, _ = more_dot.split('.', 1); print(res)
foo
>>> res, _, _ = no_dot.partition('.'); print(res)
foo
>>> res, _, _ = one_dot.partition('.'); print(res)
foo
>>> res, _, _ = more_dot.partition('.'); print(res)
foo
```

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -142,52 +139,75 @@ def task__build_repositories() -> Iterator[types.TaskDict]:
+)
+
+# All packages not referenced in `TO_BUILD` but listed in `versions.PACKAGES`
+# is supposed to be downloaded.
```
`are supposed`

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -142,52 +139,75 @@ def task__build_repositories() -> Iterator[types.TaskDict]:
 }
 
+_TO_BUILD_PKG_NAMES : FrozenSet[str] = frozenset(
+    pkg.name for pkgs in TO_BUILD.values() for pkg in pkgs
```
nested loop in comprehension are not very readable :S

gdemonet replied:
Oh, really? I like it actually, but if you prefer I can do something like:

```python
_TO_BUILD_PKG_NAMES : List[str] = []

for pkgs in TO_BUILD.values():
    for pkg in pkgs:
        _TO_BUILD_PKG_NAMES.append(pkg.name)
```

slaperche-scality replied:
Or maybe split the line as

```python
_TO_BUILD_PKG_NAMES : FrozenSet[str] = frozenset(
    pkg.name
    for pkgs in TO_BUILD.values()
    for pkg in pkgs
)
```
It's a bit more readable like that.

slaperche-scality commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -29,8 +29,8 @@
 import importlib
-import sys
 from pathlib import Path
+import sys
```
About commit message: why JSON?

I guess it's because you don't need an extra-deps (unlike YAML), but it can be good to mention if someone does archeology and ever wonder why JSON was chosen instead of YAML.

slaperche-scality commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -180,6 +180,24 @@ def _run(self) -> None:
         file_dep=[versions.VERSION_FILE],
     ),
 
+    targets.DataFile(
```
`SerializedObject` (or `SerializedData`) instead of `DataFile`.

slaperche-scality commented on buildchain/buildchain/targets/data_file.py:
```diff
@@ -0,0 +1,79 @@
+
+
+def render_json(obj: types.JSON, filepath: Path) -> None:
+    """Serialize a dict as JSON to a given file path."""
```
`a dict`: not only according to the type signature ;)

slaperche-scality commented on buildchain/buildchain/targets/data_file.py:
```diff
@@ -0,0 +1,79 @@
+        json.dump(obj, file_obj, sort_keys=True, indent=2)
+
+
+class DataFile(base.AtomicTarget):
```
`SerializedObject` (or `SerializedData`) instead of `DataFile`.

(the docstring seems to agree with me :p)

slaperche-scality commented on buildchain/buildchain/types.py:
```diff
@@ -31,3 +31,10 @@
+#       definition would have been:
+# pylint: disable=wrong-spelling-in-comment
+#       Union[str, int, float, bool, None, Dict[str, 'JSON'], List['JSON']]
+JSON = Union[str, int, float, bool, None, Dict[str, Any], List[Any]]
```
Hum, why bother?

I would go for `Any`, ok it's not accurate but it's simpler.

I mean, we already can't accurately encode the type "JSON-encodable", so let's keep things simple :)

And even now we can have `NaN` in `float` and this is not supported by JSON (Python allows it by default though)

gdemonet replied:
I'll go with `Any`

slaperche-scality commented on buildchain/buildchain/targets/data_file.py:
```diff
@@ -0,0 +1,79 @@
+        self._data = data
+        self._dest = destination
+
+        assert renderer in self.RENDERERS, (
```
An enum would be better than a string here, the set of supported rendered is known, finite and small.

slaperche-scality reviewed (changes requested):
Posting that before I forget…

I mainly reviewed the buildchain, also took a glance at Salt stuff but `Fix package manager listing of requires` would certainly benefit from a deeper review.

gdemonet committed:
Fix return types for task callables

Type hints were accurate w.r.t. the expected behaviour, but the actual
behaviour wasn't.
Note that we disable the pylint rule that complains about explicit
`return None` statements.

gdemonet committed:
Clean remnants of previous archiving method

We used to archive image tarballs with Skopeo, using the
`docker-archive` format, which wasn't working with `containerd`.
When we changed this approach to use Docker instead, we forgot to remove
some parts of the code, which we do in this commit.

See: #1250 #1253

gdemonet committed:
Clean-up Docker terms in buildchain

We attempt to provide a more obvious API by using accurate names and
concepts from the Docker concepts. In particular, our previous use of
"registry" actually meant "repository". We also make sure that image
"tags" actually represent both the name and version.

Fixes: #947

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -15,64 +14,133 @@ def __virtual__():
+    later `pkg.installed` state along with the original package.
+
+
+
```
Maybe too many empty line here :stuck_out_tongue:

TeddyAndrieux reviewed (dismissed):
LGTM: Only one minor comment

slaperche-scality commented on .pylintrc:
```diff
@@ -42,11 +42,12 @@ confidence=
+        bad-continuation,    # I kindly disagree :)
+        locally-disabled,    # Sometime we need to disable a warning locally.
+        suppressed-message,  # Don't pollute the output with useless info.
+        useless-return       # Disagreement with mypy, see https://github.com/python/mypy/issues/3974
```
Going back to that, I still think it's not needed.

I think we may have messed up our type annotations.

Seeing the examples on this issue https://github.com/python/mypy/issues/3157, one can see that the decorated function must be annotated with what it really returns (so if our function returns nothing, the return type must be `None`), because the decorator can have its own signature.

So in our case:
- `__call__` always return `None`¹
- the decorator returns a function that return `Optional[TaskError]`

With the following diff, both `pylint` and `mypy` are happy and we don't need to disable anything.

```diff
diff --git a/.pylintrc b/.pylintrc
index ce7a5299..c4aae53d 100644
--- a/.pylintrc
+++ b/.pylintrc
@@ -47,7 +47,6 @@ disable=abstract-method,     # Too many false positive :/
         bad-continuation,    # I kindly disagree :)
         locally-disabled,    # Sometime we need to disable a warning locally.
         suppressed-message,  # Don't pollute the output with useless info.
-        useless-return       # Disagreement with mypy, see https://github.com/python/mypy/issues/3974
 
 # }}}
 # Basic {{{
diff --git a/buildchain/buildchain/docker_command.py b/buildchain/buildchain/docker_command.py
index 56e53c71..4cf26444 100644
--- a/buildchain/buildchain/docker_command.py
+++ b/buildchain/buildchain/docker_command.py
@@ -140,7 +140,7 @@ class DockerBuild:
 
     @task_error(docker.errors.BuildError, handler=build_error_handler)
     @task_error(docker.errors.APIError)
-    def __call__(self) -> Optional[TaskError]:
+    def __call__(self) -> None:
         DOCKER_CLIENT.images.build(
             tag=self.tag,
             path=self.path,
@@ -148,7 +148,6 @@ class DockerBuild:
             buildargs=self.buildargs,
             forcerm=True,
         )
-        return None
 
 
 class DockerRun:
@@ -266,14 +265,13 @@ class DockerRun:
     @task_error(docker.errors.ContainerError, handler=container_error_handler)
     @task_error(docker.errors.ImageNotFound)
     @task_error(docker.errors.APIError)
-    def __call__(self) -> Optional[TaskError]:
+    def __call__(self) -> None:
         run_config = self.expand_config()
         DOCKER_CLIENT.containers.run(
             image=self.builder.tag,
             command=self.command,
             **run_config
         )
-        return None
 
 
 class DockerTag:
@@ -294,10 +292,9 @@ class DockerTag:
 
     @task_error(docker.errors.BuildError, handler=build_error_handler)
     @task_error(docker.errors.APIError)
-    def __call__(self) -> Optional[TaskError]:
+    def __call__(self) -> None:
         to_tag = DOCKER_CLIENT.images.get(self.full_name)
         to_tag.tag(self.repository, tag=self.version)
-        return None
 
 
 class DockerPull:
@@ -338,7 +335,6 @@ class DockerPull:
                     s=self, observed_digest=pulled.id
                 )
             )
-
         return None
 
 
@@ -358,12 +354,10 @@ class DockerSave:
 
     @task_error(docker.errors.APIError)
     @task_error(OSError)
-    def __call__(self) -> Optional[TaskError]:
+    def __call__(self) -> None:
         to_save = DOCKER_CLIENT.images.get(self.tag)
         image_stream = to_save.save(named=True)
 
         with self.save_path.open('wb') as image_file:
             for chunk in image_stream:
                 image_file.write(chunk)
-
-        return None
```

¹: except the one you modified to check the digest that explicitly returns a task error (we may want to raise an exception instead and use `task_error` for that, to be consistent).

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.2/improvement/640-single-source-of-image-versions` with contents from `w/2.1/improvement/640-single-source-of-image-versions`
and `development/2.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.2/improvement/640-single-source-of-image-versions origin/development/2.2
 $ git merge origin/w/2.1/improvement/640-single-source-of-image-versions
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.2/improvement/640-single-source-of-image-versions
```

slaperche-scality reviewed (changes requested):
LGTM except for one comment.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

gdemonet commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
